curvefs/src/volume/block_device_aio.cpp

```diff
@@ -0,0 +1,292 @@
+    os << "[off: " << aio->offset << ", len: " << aio->length
+       << ", ret: " << aio->ret << ", type: " << ToString(aio->op) << "]";
+
+    return os;
+}
+
+void AioReadCallBack(CurveAioContext* aio) {
+    AioRead* read = reinterpret_cast<AioRead*>(reinterpret_cast<char*>(aio) -
+                                               offsetof(AioRead, aio));
+
+    std::lock_guard<std::mutex> lock(read->mtx);
+    read->done = true;
+    read->cond.notify_one();
+}
+
+void AioWriteCallBack(CurveAioContext* aio) {
+    AioWrite* write = reinterpret_cast<AioWrite*>(reinterpret_cast<char*>(aio) -
+                                                  offsetof(AioWrite, aio));
+
+    std::lock_guard<std::mutex> lock(write->mtx);
+    write->done = true;
+    write->cond.notify_one();
+}
+
+void AioWritePaddingReadCallBack(CurveAioContext* aio) {
+    AioWrite::PaddingRead* padding = reinterpret_cast<AioWrite::PaddingRead*>(
+        reinterpret_cast<char*>(aio) - offsetof(AioWrite::PaddingRead, aio));
+
+    padding->base->OnPaddingReadComplete(aio);
+}
+
+}  // namespace
+
+AioRead::AioRead(off_t offset,
+                 size_t length,
+                 char* data,
+                 FileClient* dev,
+                 int fd)
+    : aio(), offset(offset), length(length), data(data), dev(dev), fd(fd) {}
+
+void AioRead::Issue() {
+    if (is_aligned(offset, IO_ALIGNED_BLOCK_SIZE) &&
+        is_aligned(length, IO_ALIGNED_BLOCK_SIZE)) {
+        aio.ret = -1;
+        aio.op = LIBCURVE_OP_READ;
+        aio.cb = AioReadCallBack;
+        aio.offset = offset;
+        aio.length = length;
+        aio.buf = data;
+
+        int ret = dev->AioRead(fd, &aio);
+        if (ret < 0) {
+            LOG(ERROR) << "Failed to issue aio read: " << &aio;
+            done = true;
+        }
+
+        return;
+    }
+
+    // unaligned read request
+    padding = absl::make_unique<Padding>();
+    auto alignedOff = align_down(offset, IO_ALIGNED_BLOCK_SIZE);
+    auto alignedEnd = align_up(offset + length, IO_ALIGNED_BLOCK_SIZE);
+
+    padding->offset = alignedOff;
+    padding->length = alignedEnd - alignedOff;
+    padding->data.reset(new char[padding->length]);
+
+    aio.ret = -1;
+    aio.op = LIBCURVE_OP_READ;
+    aio.cb = AioReadCallBack;
+    aio.offset = padding->offset;
+    aio.length = padding->length;
+    aio.buf = padding->data.get();
+
+    int ret = dev->AioRead(fd, &aio);
+    if (ret < 0) {
+        LOG(ERROR) << "Failed to issue aio read: " << &aio;
+        done = true;
+    }
+}
+
+ssize_t AioRead::Wait() {
+    std::unique_lock<std::mutex> lock(mtx);
+    cond.wait(lock, [this]() { return done; });
+
+    if (static_cast<ssize_t>(aio.ret) != static_cast<ssize_t>(aio.length)) {
+        LOG(ERROR) << "AioRead error: " << &aio;
+        return -1;
+    }
+
+    if (padding != nullptr) {
+        std::memcpy(data, padding->data.get() + (offset - padding->offset),
+                    length);
+    }
```

Comment:
padding->data.reset(new char[padding->length]);

not free

Reply:
the unique_ptr's destructor will free the memory